### PR TITLE
device: move {In,Out}boundElement Mutex to container type

### PR DESCRIPTION
**Pull request by jwhited**
{In,Out}boundElement locking can contribute to significant overhead via sync.Mutex.lockSlow() in some environments. These types are passed throughout the device package as elements in a slice, so move the per-element Mutex to a container around the slice.

### CODE REVIEW

**Review by valscale — COMMENTED** [review]
This is the first time I've looked at this code, but it looks good to me!

File: device/pools.go [hunk 1]
```diff
@@ -65,28 +65,32 @@ func (device *Device) PopulatePools() {
 	})
 }
 
-func (device *Device) GetOutboundElementsSlice() *[]*QueueOutboundElement {
-	return device.pool.outboundElementsSlice.Get().(*[]*QueueOutboundElement)
+func (device *Device) GetInboundElementsContainer() *QueueInboundElementsContainer {
+	c := device.pool.inboundElementsContainer.Get().(*QueueInboundElementsContainer)
+	c.Mutex = sync.Mutex{}
```
Comment (maisem) [inline device/pools.go:70]: This looks sus, what're you trying to do?

File: device/send.go [hunk 2]
```diff
@@ -46,18 +46,21 @@ import (
...
 
+type QueueOutboundElementsContainer struct {
+	sync.Mutex
+	elems []*QueueOutboundElement
+}
+
 func (device *Device) NewOutboundElement() *QueueOutboundElement {
 	elem := device.GetOutboundElement()
 	elem.buffer = device.GetMessageBuffer()
-	elem.Mutex = sync.Mutex{}
```
Comment (maisem) [inline device/send.go:60]: Oh this is so weird

File: device/pools.go (hunk 1 above)
Comment (maisem) [reply device/pools.go:70]: I think I've convinced myself that the locking is just plain weird here, and that you aren't making it worse.

File: device/pools.go [hunk 3]
```diff
@@ -46,13 +46,13 @@ func (p *WaitPool) Put(x any) {
...
 func (device *Device) PopulatePools() {
-	device.pool.outboundElementsSlice = NewWaitPool(PreallocatedBuffersPerPool, func() any {
-		s := make([]*QueueOutboundElement, 0, device.BatchSize())
-		return &s
-	})
-	device.pool.inboundElementsSlice = NewWaitPool(PreallocatedBuffersPerPool, func() any {
+	device.pool.inboundElementsContainer = NewWaitPool(PreallocatedBuffersPerPool, func() any {
 		s := make([]*QueueInboundElement, 0, device.BatchSize())
-		return &s
+		return &QueueInboundElementsContainer{elems: s}
```
Comment (maisem) [inline device/pools.go:51]: you can inline this now

**Review by maisem — APPROVED** [review]

**Pull request merged**